library/std/src/Std/StatePreparation.qs

```diff
@@ -351,7 +352,8 @@ operation ApproximatelyMultiplexZ(
         ApproximatelyMultiplexZ(tolerance, coefficients0, control, target);
         if AnyOutsideToleranceD(tolerance, coefficients1) {
             within {
-                Controlled X(controlRegister, target);
+                // Reduce circuit complexity by using aux qubits.
+                Controlled ApplyAsSinglyControlled(controlRegister, (X, target));
```

Comment:
One can still save a factor of 2x by only doing the forward direction of the multi-controlled X gate (chain of computing ANDs) and then the backward direction (chain of uncomputing ANDs) afterwards. However, this will require to allocate the helper qubits in this operation.

For a 3-qubit controlRegister the trace would be as follows:

```qsharp
use helper = Qubit[1];

And(controlRegister[0], controlRegister[1], helper[0]);
And(controlRegister[1], helper[0], target);

ApproximatelyMultiplexZ(tolerance, coefficients1, control, target);

Adjoint And(controlRegister[1], helper[0], target);
Adjoint And(controlRegister[0], controlRegister[1], helper[0]);
```